File: tests/fixtures/process_stdio_client.mjs
import { Client } from '@modelcontextprotocol/sdk/client/index.js';
import { StdioClientTransport } from '@modelcontextprotocol/sdk/client/stdio.js';

const serverPath = process.argv[2];
if (!serverPath) {
  throw new Error('usage: node process_stdio_client.mjs <server-executable>');
}

const client = new Client({
  name: 'cxxmcp-ts-process-client',
  version: '1.0.0',
});

const transport = new StdioClientTransport({
  command: serverPath,
  args: [],
});

await client.connect(transport);

try {
  const tools = await client.listTools();
  if (!tools.tools.some((tool) => tool.name === 'echo')) {
    throw new Error(`echo tool not found: ${JSON.stringify(tools)}`);
  }

  const result = await client.callTool({
    name: 'echo',
    arguments: { value: 42 },
  });
  const text = result.content?.[0]?.text;
  if (text !== 'echo') {
    throw new Error(`unexpected echo result: ${JSON.stringify(result)}`);
  }
  if (result.structuredContent?.value !== 42) {
    throw new Error(`unexpected structured content: ${JSON.stringify(result)}`);
  }
} finally {
  await client.close();
}

File: tests/fixtures/process_stdio_client.py
import asyncio
import importlib.metadata
import sys

from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client


async def main() -> None:
    if len(sys.argv) != 2:
        raise RuntimeError("usage: python process_stdio_client.py <server-executable>")

    print(f"using mcp=={importlib.metadata.version('mcp')}")

    params = StdioServerParameters(command=sys.argv[1], args=[])
    async with stdio_client(params) as streams:
        read, write = streams
        async with ClientSession(read, write) as session:
            await session.initialize()

            tools = await session.list_tools()
            if not any(tool.name == "echo" for tool in tools.tools):
                raise RuntimeError(f"echo tool not found: {tools!r}")

            result = await session.call_tool("echo", arguments={"value": 42})
            text = getattr(result.content[0], "text", None) if result.content else None
            if text != "echo":
                raise RuntimeError(f"unexpected echo result: {result!r}")


if __name__ == "__main__":
    asyncio.run(main())
